File: problems/C++/90. Subsets II.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> path;
        bt(0, path, nums, res);
        return res;
    }

private:
    void bt(int start, vector<int>& path, vector<int>& nums, vector<vector<int>>& res) {
        res.push_back(path);
        for (int i = start; i < nums.size(); ++i) {
            if (i > start && nums[i] == nums[i-1]) {
                continue;
            }
            path.push_back(nums[i]);
            bt(i + 1, path, nums, res);
            path.pop_back();
        }
    }
};

File: problems/Java/90. Subsets II.java
class Solution {

    public List<List<Integer>> subsetsWithDup(int[] nums) {

        Arrays.sort(nums);

        List<List<Integer>> res = new ArrayList<>();

        bt(0, new ArrayList<>(), nums, res);

        return res;

    }

    private void bt(int start, List<Integer> path, int[] nums, List<List<Integer>> res) {

        res.add(new ArrayList<>(path));

        for (int i = start; i < nums.length; i++) {

            if (i > start && nums[i] == nums[i-1]) {

                continue;

            }

            path.add(nums[i]);

            bt(i + 1, path, nums, res);

            path.remove(path.size() - 1);

        }

    }

}